scripts/globals/npc_util.lua

```diff
@@ -38,175 +49,254 @@ function npcUtil.pickNewPosition(npc, positionTable, allowCurrentPosition)
+    player:completeQuest(area,quest);
     return true
 end
 
--- returns true if the trade conditions are met
--- npcUtil.tradeHas(trade, 532)
--- npcUtil.tradeHas(trade, {532, 927})
--- npcUtil.tradeHas(trade, nil, 5)
-function npcUtil.tradeHas(trade, items, gil)
-    local itemCount = {};
-
-    if (gil ~= nil and trade:getGil() ~= gil) then
-        -- given wrong armound of gil
-        return false;
-    elseif (trade:getGil() ~= 0 and gil == nil) then
-        -- given gil when I don't want it
-            return false;
+--[[ *******************************************************************************
+    check whether trade has all required items
+        if yes, confirm all the items and return true
+        if no, return false
+        
+    valid examples of items:
+        640                     -- copper ore x1
+        { 640, 641 }            -- copper ore x1, tin ore x1
+        { 640, 640 }            -- copper ore x2
+        { {640,2} }             -- copper ore x2
+        { {640,2}, 641 }        -- copper ore x2, tin ore x1
+        { 640, {"gil", 200} }   -- copper ore x1, gil x200
+******************************************************************************* --]]
+function npcUtil.tradeHas(trade, items, exact)
+    if (type(exact) ~= "boolean") then exact = false; end
+
+    -- create table of traded items, with key/val of itemId/itemQty
+    local tradedItems = {};
+    local itemId;
+    local itemQty;
+    for i = 0, trade:getSlotCount()-1 do
+        itemId = trade:getItemId(i);
+        itemQty = trade:getItemQty(itemId);
+        tradedItems[itemId] = itemQty;
     end
-
-    if (items ~= nil) then
-
-        if (type(items) == "number") then
-            items = {items}
-        end
-
-        -- given different amount of items
-        if (trade:getItemCount() ~= #items) then
-            return false;
-        end
-
+    
+    -- create table of needed items, with key/val of itemId/itemQty
+    local neededItems = {};
+    if (type(items) == "number") then
+        neededItems[items] = 1;
+    elseif (type(items) == "table") then
+        local itemId;
+        local itemQty;
         for _, v in pairs(items) do
-            if (itemCount[v] == nil) then
-                itemCount[v] = 1;
+            if (type(v) == "number") then
+                itemId = v;
+                itemQty = 1;
+            elseif (type(v) == "table" and #v == 2 and type(v[1]) == "number" and type(v[2]) == "number") then
+                itemId = v[1];
+                itemQty = v[2];
+            elseif (type(v) == "table" and #v == 2 and type(v[1]) == "string" and type(v[2]) == "number" and string.lower(v[1]) == "gil") then
+                itemId = 65535;
+                itemQty = v[2];
             else
-                itemCount[v] = itemCount[v] + 1
+                print("ERROR: invalid value contained within items parameter given to npcUtil.tradeHas.");
+                goto continue;
```

Comment:
instead of a goto.. can you just wrap the neededItems[itemId] line with itemId ~= nil and itemQty ~= nil?

I really don't like gotos

Reply:
Done.